### DISCUSSION TITLE
metrics

### DISCUSSION THREAD

**Discussion by gzt5142** [hytest-org/hytest#346] [Model Evaluation Methods]
Currently, HyTEST is coding statistical metrics as functions in [notebooks](https://hytest-org.github.io/hytest/evaluation/metrics.html), and these notebooks are called into [other notebooks](https://hytest-org.github.io/hytest/evaluation/tutorials/nwm_streamflow/02_Analysis_StdSuite.html#step-3-define-metrics) for use in various workflows. Using metrics in notebooks is a useful way to 'import' common functions so that we don't duplicate code all over the place.  Sourcing those common functions using a `%run` magic seems to work for us in a tutorial context, and it lets us put good descriptions (in markdown) along side the code cells (including equations and references/citations).  And this publishes well in the jupyterbook (IMO).  

At some point, you may want to consider building a stand-alone python package for metrics that you want to use *outside* of the tutorial context. We talked about this a loooooooonnnnng time ago ("pydroGOF" as a python replacement/port of the `hydroGOF` R package's functionality).

I think it's worth considering as we put more and more common functions into shared space to economize on code re-use. See the scikit-learn `metrics` sub-package as an example of how the ML folks are doing something similar.  If it is well-designed, the `hytest.metrics` module would give a hydro-specific flair to what is otherwise generic computations. 

This would be a good fit for the "domain-specific" layer in Jim Bednar's SOSA model: https://github.com/panstacks/pandata

We want to find the best way forward, whether that is contributing to another open-source python hydrological stats package or something like suggested above.

**Comment by jamaa** [comment]
You might also want to consider using or building on an existing package such as [hydroeval](https://github.com/ThibHlln/hydroeval) or [hydroerr](https://github.com/BYU-Hydroinformatics/HydroErr)

**Reply by sfoks** [reply]
I like this idea of adding back to a published package, it looks like the last commits to these packages were several + years ago. We may need to shop around to find the right fit but I think your suggestion is great. Thanks for posting!

**Comment by ThomasMGeo** [comment]
If you can adhere to the scikit-learn interface where possible, and maybe use units via [Pint ](https://pypi.org/project/Pint/) would be great!
